Added read_data functionality

saakethdesai opened the pull request:
Closes #6

saakethdesai committed:
Added read_data.h (supports atom_style "atomic" and "charge"). Also added MINIMD license

saakethdesai committed:
Make source compatible with read_data.h

streeve commented on src/cabanamd.cpp:
```diff
@@ -1,50 +1,50 @@
 /****************************************************************************
- * Copyright (c) 2018-2019 by the Cabana authors                            *
```
Revert spacing

streeve commented on src/input.cpp:
```diff
@@ -402,11 +417,10 @@ void Input::check_lammps_command(int line) {
       force_line = line;
     }
-    if(system->do_print && !known)
-      printf("LAMMPS-Command: 'pair_style' command only supports 'lj/cut' style in CabanaMD\n");
```
Add back error check

streeve commented on src/input.cpp:
```diff
@@ -402,11 +417,10 @@ void Input::check_lammps_command(int line) {
   }
   if(strcmp(input_data.words[line][0],"pair_coeff")==0) {
     known = true;
+    force_cutoff = atof(input_data.words[line][3]);
```
Remove (only for NN, in later commit)

streeve commented on src/read_data.h:
```diff
@@ -0,0 +1,230 @@
+void read_lammps_velocities(ifstream &file, System* s)
+{
+  string line;
+  auto x = Cabana::slice<Positions>(s->xvf);
```
Should only need v here (try compiling with -Wall -Wextra -pedantic -Werror and see what you get)

streeve commented on src/read_data.h:
```diff
@@ -0,0 +1,230 @@
+{
+  string line;
+  auto x = Cabana::slice<Positions>(s->xvf);
+  auto v = Cabana::slice<Velocities>(s->xvf);
```
Only slice what you need

streeve commented on src/cabanamd.cpp:
```diff
@@ -55,423 +55,428 @@
-  }
-  // Create Communication Submodule
-  comm = new Comm(system, neigh_cutoff);
+//don't understand this
```
Remove

saakethdesai committed:
Fixed spacing and added minor change requests

saakethdesai committed:
Fixed spacing and removed unneccessary slicing

streeve commented on src/cabanamd.cpp:
```diff
@@ -120,6 +123,8 @@ void CabanaMD::init(int argc, char* argv[]) {
   // Compute NeighList
   force->create_neigh_list(system);
 
+  //Do some additional settings
```
Remove?

streeve commented on src/cabanamd.cpp:
```diff
@@ -255,39 +260,39 @@ void CabanaMD::run(int nsteps) {
-  double time = timer.seconds();
+}
 
-  if(system->do_print) {
```
Check spacing one more time

streeve commented on src/read_data.h:
```diff
@@ -0,0 +1,220 @@
+  s->resize(s->N);
+  
+  //perform domain decomposition and get access to subdomains
+  //comm->create_domain_decomposition(); //TODO: Look into this 
```
Uncomment (and check if added somewhere else)

streeve commented on src/read_data.h:
```diff
@@ -0,0 +1,220 @@
+  // proc 0 reads upto non-blank line plus 1 following line
+  // eof is set to 1 if any read hits end-of-file
+  string line, keyword;
+  while(1) {
```
Exit if end of file

streeve commented on src/read_data.h:
```diff
@@ -0,0 +1,220 @@
+  
+  T_INT id_tmp, type_tmp;
+  T_FLOAT x_tmp, y_tmp, z_tmp, q_tmp; 
+  for (int n=0; n < s->N; n++) {
```
Check that the current atom is within the subdomain

streeve commented on src/read_data.h:
```diff
@@ -0,0 +1,220 @@
+    }
+    else if(line.find("xlo xhi") != string::npos) {
+      sscanf(temp, "%lg %lg", &xlo, &xhi);
+      s->domain_x = xhi - xlo;
```
Read box lo (or add later?)

saakethdesai committed:
Added box lo and box hi reading capabilities. Pushed domain
decomposition to read_data.h and pushed read_data to after declaring
comm pointer

streeve commented on src/cabanamd.cpp:
```diff
@@ -107,6 +111,8 @@ void CabanaMD::init(int argc, char* argv[]) {
   // Ok lets go ahead and create the particles if that didn't happen yet
   if(system->N == 0)
     input->create_lattice(comm);
+  if(system->N == 0 && input->read_data_flag==true)
```
Shouldn't this be read_data OR create_lattice ?

streeve commented on LICENSE_MINIMD:
```diff
@@ -0,0 +1,167 @@
+		   GNU LESSER GENERAL PUBLIC LICENSE
```
Should add the Cabana brief license on top

streeve replied:
My mistake (meant to suggest read_data, as you fixed)

saakethdesai committed:
Added Cabana license on top of MiniMD license. Fixed logic of reading
data or creating a lattice

saakethdesai committed:
Re-added Cabana license to MiniMD license file

saakethdesai committed:
Added license to read_data.h

saakethdesai committed:
Updating license file

saakethdesai committed:
Updating MiniMD license file

streeve commented on src/cabanamd.cpp:
```diff
@@ -105,7 +109,9 @@ void CabanaMD::init(int argc, char* argv[]) {
-  if(system->N == 0)
+  if(system->N == 0 && input->read_data_flag==true)
+    read_lammps_data_file(input->lammps_data_file, system, comm);
+  else if(system->N == 0 && input->read_data_flag=false)
```
Remove the redundant flag check

streeve commented on src/read_data.h:
```diff
@@ -0,0 +1,255 @@
+      continue;
+    
+  }
+}
```
Return string if no early exit (Werror)

streeve reviewed (changes requested):
Check with a generic lammps data file to confirm everything is working as intended

saakethdesai committed:
Added error checking for invalid files or file paths > 32 characters

saakethdesai committed:
Error out if Masses are found in data file

saakethdesai committed:
Fixed else block brackets (error from flags)
